### ISSUE TITLE
NSJS functions do not clean up when they fail?

### ISSUE THREAD

**Issue by LoganDark** [danielyxie/bitburner#464]
![image](https://user-images.githubusercontent.com/4723091/45687784-7a3ecf00-bb04-11e8-9ab1-b1aca06417e2.png)

I have a loop in my JS script that looks for servers I don't have root access to and attempts to hack them.

![image](https://user-images.githubusercontent.com/4723091/45687824-9b072480-bb04-11e8-8f62-c6163ab75a7e.png)

Nice...

`hack` (awaited, of course) is in a try/catch, so when it fails, the script prints an error message. Nice.

But during the next iteration, when it tries to check if we have root access for the next IP, it fails because Bitburner still thinks `hack` is still running.

This is an issue, since I'd like to be able to continue my loop without restarting the script if something fails.

**Comment by LoganDark** [comment]
The only solution is to exit the entire script... cleanly. `ns.exit()` doesn't work because it's "concurrent" as well, so I have to use a label and use `break mainloop`.

**Comment by danielyxie** [comment]
Could you post the script? Or the section of the script that causes this issue?

**Comment by LoganDark** [comment]
```js
try {
	await ns.hack('hostname')
} catch (e) {
	await ns.tprint('Error: ' + e.toString())
}
```

I haven't tested it, but that's the concept - it will cause a concurrency error if the hack fails.

**Comment by danielyxie** [comment]
I put in a patch for this on the [beta branch](https://www.reddit.com/r/Bitburner/comments/9f9lwh/beta_branch/).

**Comment by danielyxie** [comment]
Fixed in v0.41.0

**Comment by LoganDark** [comment]
Ty~
